resteasy-jaxrs/src/main/java/org/jboss/resteasy/core/ContextParameterInjector.java

```diff
@@ -25,31 +30,62 @@
 @SuppressWarnings("unchecked")
 public class ContextParameterInjector implements ValueInjector
 {
-   private Class type;
+   private Class rawType;
    private Class proxy;
    private ResteasyProviderFactory factory;
+   private Type genericType;
+   private Annotation[] annotations;
 
-   public ContextParameterInjector(Class proxy, Class type, ResteasyProviderFactory factory)
+   public ContextParameterInjector(Class proxy, Class rawType, Type genericType, Annotation[] annotations, ResteasyProviderFactory factory)
    {
-      this.type = type;
+      this.rawType = rawType;
+      this.genericType = genericType;
       this.proxy = proxy;
       this.factory = factory;
+      this.annotations = annotations;
    }
 
-   public Object inject(HttpRequest request, HttpResponse response)
+   @Override
+   public CompletionStage<Object> inject(HttpRequest request, HttpResponse response, boolean unwrapAsync)
    {
       // we always inject a proxy for interface types just in case the per-request target is a pooled object
       // i.e. in the case of an SLSB
-      if (type.equals(Providers.class)) return factory;
-      if (!type.isInterface() || type.equals(SseEventSink.class))
+      if (rawType.equals(Providers.class)) return CompletableFuture.completedFuture(factory);
+      if (!rawType.isInterface() || rawType.equals(SseEventSink.class) || factory.hasAsyncContextData(genericType))
       {
-         return ResteasyProviderFactory.getContextData(type);
+         return unwrapIfRequired(request, factory.getContextData(rawType, genericType, annotations, unwrapAsync), unwrapAsync);
       }
-      else if (type.equals(Sse.class))
+      else if (rawType.equals(Sse.class))
       {
-         return new SseImpl();
+         return CompletableFuture.completedFuture(new SseImpl());
       }
-      return createProxy();
+      return CompletableFuture.completedFuture(createProxy());
+   }
+
+   private CompletionStage<Object> unwrapIfRequired(HttpRequest request, Object contextData, boolean unwrapAsync)
+   {
+      if(unwrapAsync && rawType != CompletionStage.class && contextData instanceof CompletionStage) {
+         // FIXME: do not unwrap if we have no request?
```

Comment:
What about this? What is the plan for the future?

Reply:
No, that is outdated, I fixed that question with the `unwrapAsync` param. I'll remove it.